File: flutter/android/app/src/main/kotlin/com/infobip/infobip_rtc_flutter_showcase/video/RTCVideoPlugin.kt
package com.infobip.infobip_rtc_flutter_showcase.video

import com.infobip.webrtc.sdk.api.InfobipRTC
import com.infobip.webrtc.sdk.api.call.WebrtcCall
import com.infobip.webrtc.sdk.api.model.video.RTCVideoTrack
import io.flutter.embedding.engine.plugins.FlutterPlugin
import java.util.*

class RTCVideoPlugin: FlutterPlugin {
  private val views: MutableMap<String, RTCVideoView> = HashMap()

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    flutterPluginBinding.platformViewRegistry
      .registerViewFactory(
        "<infobip-rtc-video-view>",
        RTCVideoViewFactory(this)
      )
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
  }

  fun setTrack(streamId: String, track: RTCVideoTrack?) {
    views[streamId]?.let { view ->
      if (track == null) {
        view.release()
      } else {
        view.attachTrack(track)
      }
    }
  }

  fun setView(streamId: String, rtcVideoView: RTCVideoView) {
    views[streamId] = rtcVideoView
    val activeCall = InfobipRTC.getInstance().activeCall as? WebrtcCall

    if (streamId == "remote") {
      rtcVideoView.attachTrack(activeCall?.remoteCameraTrack())
    } else {
      rtcVideoView.attachTrack(activeCall?.localCameraTrack())
    }
  }
}

File: flutter/ios/Runner/video/RTCVideoPlugin.swift
import InfobipRTC

public class RTCVideoPlugin: NSObject, FlutterPlugin {
    var videoViews: [String: RTCVideoView] = [:]

    public static func register(with registrar: FlutterPluginRegistrar) {
        registrar.register(
            RTCVideoViewFactory(messenger: registrar.messenger(), plugin: RTCVideoPlugin()),
            withId: "<infobip-rtc-video-view>"
        )
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(FlutterMethodNotImplemented)
    }

    func setView(streamId: String, rtcVideoView: RTCVideoView) {
        videoViews[streamId] = rtcVideoView
        if let activeCall = getInfobipRTCInstance().getActiveCall() as? WebrtcCall {
            if streamId == "remote", let track = activeCall.remoteCameraTrack() {
                rtcVideoView.attachTrack(track)
            } else if streamId == "local", let track = activeCall.localCameraTrack() {
                rtcVideoView.attachTrack(track)
            }
        }
    }

    func setTrack(streamId: String, track: VideoTrack?) {
        if let view = videoViews[streamId] {
            if track == nil {
                view.release()
            } else {
                view.attachTrack(track)
            }
        }
    }

    func clearViews() {
        videoViews.values.forEach { $0.release() }
        videoViews.removeAll()
    }
}
